### PR TITLE
[AzureResourceManagerTemplateDeploymentV3] [AzureResourceGroupDeploymentV2] Deployment outputs fixed and updated documentation

### LINKED ISSUE #16280: Setting output variable directly from AzureResourceManagerTemplateDeployment@3 deployment output value gives leading whitespace
## Required Information

**Question, Bug, or Feature?**  
*Type*: Bug

**Enter Task Name**: AzureResourceManagerTemplateDeployment@3

## Environment
- Server - Azure Pipelines
- Agent - Private, Ubuntu20.04 
  - Azure DevOps + VMSS
  - VM-extension (Microsoft.Azure.DevOps.Pipelines.Agent - Microsoft.VisualStudio.Services.TeamServicesAgentLinux - 1.22)
    
## Issue Description

Setting output variables with "##vso[task.setvariable]" using the deployment output values directly from AzureResourceManagerTemplateDeployment@3 task gives leading whitespace in the results.

image

```
- task: AzureResourceManagerTemplateDeployment@3
  displayName: Deploy ExpressRoute Circuit
  inputs:
    <obfuscated>
    deploymentMode: Incremental
    deploymentOutputs: ercOutputs
```

```
- powershell: |
    Write-Host "##vso[task.setvariable variable=ercId;isOutput=true]$(ercOutputs.ercId.value)"
    Write-Host "##vso[task.setvariable variable=ercLocation;isOutput=true]$(ercOutputs.ercLocation.value)"
  displayName: 'Write output variables'
  name: ercVars

```

Next job using the output variables as stageDependencies:

```
- stage: MyNextStage
  dependsOn:
  - TheOtherStage

  jobs:

  - job: ExpressRouteGatewayAndConnection
    displayName: 'Deploy Expressroute Gateway and Connection'
    variables:
      ercId: $[stageDependencies.ExpressRouteCircuit.DeployExpressRouteCircuit.outputs['ercVars.ercId']]
      ercLocation: $[stageDependencies.ExpressRouteCircuit.DeployExpressRouteCircuit.outputs['ercVars.ercLocation']]
```

### Task logs

The job preparation parameters result always have a leading whitespace, and I guess this is not intended. This seems to have made the parameter validation for my next job fail atleast.

image

### Troubleshooting

I am fully aware that I can solve this by doing the traditional example.

```
        $json = ConvertFrom-Json '$(ercOutputs)'
        Write-Host "##vso[task.setvariable variable=ercId;isOutput=true]$($json.ercId.value)"
        Write-Host "##vso[task.setvariable variable=ercLocation;isOutput=true]$($json.ercLocation.value)"
```

However if this could be used properly it would feel atleast a bit smoother working with the outputs. 

### Error logs

```
2022-05-04T14:38:29.1504100Z There were errors in your deployment. Error code: InvalidDeploymentParameterKey.
2022-05-04T14:38:29.1507327Z ##[error]One of the deployment parameters has an empty key. Please see https://aka.ms/resource-manager-parameter-files for details.
2022-05-04T14:38:29.1528085Z ##[error]Check out the troubleshooting guide to see if your issue is addressed: https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/deploy/azure-resource-group-deployment?view=azure-devops#troubleshooting
2022-05-04T14:38:29.1542329Z ##[error]Task failed while creating or updating the template deployment.
```

### PR DESCRIPTION
**Pull request by svrooij**
**Task name**: AzureResourceManagerTemplateDeploymentV3 and AzureResourceGroupDeploymentV2

**Description**: Outputs had quotes around them, which made them useless. According to [about task.setVariable](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/set-variables-scripts?view=azure-devops&tabs=bash#about-tasksetvariable) the added quotes shouldn't be there. Having direct access to these variables without the need to write your own script to parse, was not in the documentation.

**Documentation changes required:** Y

**Added unit tests:** N (I'm unable to run the tests on my machine)

**Attached related issue:** Y

**Checklist**:
- [X] Task version was bumped - please check [instruction](https://github.com/microsoft/azure-pipelines-tasks/tree/master/docs/taskversionbumping.md) how to do it
- [X] Checked that applied changes work as expected

Related https://github.com/MicrosoftDocs/azure-devops-docs/issues/12277
Fixed #16280

### CODE REVIEW

**Comment by svrooij** [comment]
It's unclear to me, why the output parsing switched from `tl.setVariable(``${path}.${key}``, JSON.stringify(obj[key]));` to the console output way (which does the same). But since this feature was there in November 2020 the output hasn't been filled correctly. It wants a string value, not a JSON encoded value of a variable.

## Issue with JSON.stringify
I have this ARM template that has an output called `functionName`, and I want to use it to deploy a function. Before this fix the `armOutputs.functionName.value` would contain `"my-new-azure-functions-app"` and if I try to use that to deploy my new function app I'm getting an error stating `Resource '"my-new-azure-functions-app"' doesn't exist.` Which is true, since the function app doesn't have the `"` in the name. A work around it to use `$(replace(armOutputs.functionName.value, '"', ''))` but that just seems ugly.

The `tl.setVariable(...)` eventually uses [this code](https://github.com/microsoft/azure-pipelines-task-lib/blob/5ebf55f67908c5b16314ccb60a0c6b94dacd8a0a/node/taskcommand.ts#L26-L51) where is turns any value to a string by doing `let message: string = '' + (this.message || '');`.

**Comment by svrooij** [comment]
Looks like someone else also found an issue with the outputs, the same day I found mine.... #16280 They must be related and probably fixed in this PR.

**Comment by svrooij** [comment]
@rohit-batra do you have some insights on how you guys process pull requests?

**Comment by svrooij** [comment]
It seems these tasks are no longer "owned" by someone? Been waiting for almost 3 weeks to even get any response from the code owner.
Can you @v-nagarajku or @anatolybolshakov help me with getting this some attention or getting this merged?

**Comment by svrooij** [comment]
@max-zaytsev since you where the last to commit to one of the modules I'm changing, can you have a look at this PR? Or point me in the right direction on how to get this pr some attention?

**Comment by svrooij** [comment]
@AndreyIvanov42 who do I ping to get this small pr some more attention? The code owner seems to be unresponsive.

which tree should I shake? I see a lot of pull requests staying unanswered for weeks.

**Comment by feliasson** [comment]
Hi again @svrooij, sad to see they have not yet looked at your PR...

However, what would you think about [this](https://github.com/microsoft/azure-pipelines-tasks/issues/16468) in addition to your proposed changes here as well ? 

Just sample [commit](https://github.com/microsoft/azure-pipelines-tasks/compare/master...feliasson:master), sadly a bit unsure how to test the tasks after the changes myself :)

**Comment by svrooij** [comment]
I was happy to finally see a response to my pr... Then I saw you actually want to add more to my pr.

I got a feeling they are not really actively monitoring this repository.

On the set output part, what does it do? Maybe it should be a switch? People should not use outputs for things that should be secret but I guess some do. Having that information available to all steps might be a security issue?

**Comment by feliasson** [comment]
> I was happy to finally see a response to my pr... Then I saw you actually want to add more to my pr.
> 
> I got a feeling they are not really actively monitoring this repository.
> 
> On the set output part, what does it do? Maybe it should be a switch? People should not use outputs for things that should be secret but I guess some do. Having that information available to all steps might be a security issue?

It enables the output to be used in [other stages/jobs](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/variables?view=azure-devops&tabs=yaml%2Cbatch#set-a-multi-job-output-variable) for example, A switch would be even better, right now you have to create a small task, assign it a name, and set the variables again after every deployment task you have done, i.e.
```
- job: A
  
  steps:
  - task: AzureResourceManagerTemplateDeployment@3
    name: DeployNetworkingTask
    <---->
    deploymentOutputs: networkOutputs 
```
```
  - pwsh: |
      $outputs = ConvertFrom-Json '$(networkOutputs)' -Depth 20
      Write-Host "##vso[task.setvariable variable=networks;isOutput=true]$($outputs.networks.value | ConvertTo-Json -Depth 20 -Compress)"
    name: networkVars
```
```
- job: B
  dependsOn: A
  variables:
    networks: $[ dependencies.A.outputs['networkVars.networkOutputs'] ]
```

Instead of being able to directly:
```
- job: A
  
  steps:
  - task: AzureResourceManagerTemplateDeployment@3
    name: DeployNetworkingTask
    <---->
    deploymentOutputs: networkOutputs 
```

```
- job: B
  dependsOn: A
  variables:
    networks: $[ dependencies.A.outputs['DeployNetworkingTask.networkOutputs'] ]
```
Adding a way to have the output set with `isOutput=true` would enable this, making it much smoother,

**Comment by svrooij** [comment]
@feliasson this is a great idea, but I think that should be switchable. Not everybody would want that enabled by default.
This PR fixes some functionality, I hoped that would make it easier to get it approved.

In great suggestion, introduces new functionality and might make it had to get approved. With this pr approved, setting those variables as output variables would become easier.

**Comment by svrooij** [comment]
In my last effort to get this pr some attention I'm hoping one of the following people might be able to help, the code owner is unreachable. Maybe even left Microsoft? This PR is hanging for more then 2 months now...
@max-zaytsev @DenisRumyantsev @vineetmimrot @ddivanshu

**Comment by svrooij** [comment]
@max-zaytsev all the tests succeeded. What should happen next?

**Comment by svrooij** [comment]
@max-zaytsev or @WilliamAntonRohm  or @rvairavelu how do we proceed with this PR? There doesn't seem to be any change on this issue.

How do I publish my own version so I can use it in Azure DevOps, I don't want to wait indefinitely without any feedback what so ever.

**Review by rajnish-byte — APPROVED** [review]
Approving on behalf of Renuka

**Comment by svrooij** [comment]
@rajnish-byte can you also start the pipeline? Do you want me to merge the latest version?

And I've seen that I updated the `package-lock.json` to lockFileVersion 2, do I need to revert that change?

**Comment by rajnish-byte** [comment]
@svrooij started the pipelines. Not sure of the impacts of the version change here
@rvairavelu

**Comment by svrooij** [comment]
@rajnish-byte it seems to have failed on windows due to a timeout. How do I fix this?

Locally (Windows 11) that specific tasks takes 5 seconds, which is pretty long for what it is testing.

**Comment by svrooij** [comment]
Having tests fail because it cannot download a file from Github 😢 

image

**Comment by svrooij** [comment]
Yeah thanks bot..... This is really helpful, not responding to a PR and then closing after 180 days.

**Comment by svrooij** [comment]
@DenisRumyantsev any change you got some time to review this PR and finally get it going?

**Comment by DenisRumyantsev** [comment]
Hi, @ozanovus, could you review this pull request, please?

Hi, @svrooij, could you resolve conflicts?

**Comment by svrooij** [comment]
@DenisRumyantsev I've just merged all the required files and bumped the task version for the 4th time. How does this proceed?

@ozanovus are you the code owner? Github says I need to talk to @rvairavelu or @manolerazvan according to the `codeowners` file

**Comment by v-nagarajku** [comment]
@svrooij - reviewing this PR. Will keep you posted if we have any comments.

**Comment by svrooij** [comment]
@v-nagarajku My node version was too "new" and created a v2 lockfile.

**Comment by v-nagarajku** [comment]
@svrooij what's your node version and which task .lock file created?

**Comment by svrooij** [comment]
@v-nagarajku i had 18.something installed. Switched back to 10.somethinf and runned `npm install`

File: Tasks/AzureResourceGroupDeploymentV2/task.json
Comment (ozanovus) [inline Tasks/AzureResourceGroupDeploymentV2/task.json:17]: First things first, let's bump it to 220 according to [versioning guide](https://github.com/microsoft/azure-pipelines-tasks/blob/master/docs/taskversionbumping.md); since we passed cut-off date.

File: Tasks/AzureResourceGroupDeploymentV2/Tests/L0.ts
Comment (ozanovus) [inline Tasks/AzureResourceGroupDeploymentV2/Tests/L0.ts:12]: Do you have a complaint for previous value of timeout? Could you please give us more details why you needed this change?

File: Tasks/AzureResourceGroupDeploymentV2/operations/ResourceGroup.ts
Comment (ozanovus) [inline Tasks/AzureResourceGroupDeploymentV2/operations/ResourceGroup.ts:566]: Could you please give us more insights for the cases that are not "array" ?
I'm trying to understand why you've limited  `JSON.stringify` operation only for `Array.isArray`.

File: Tasks/AzureResourceGroupDeploymentV2/task.json
Comment (svrooij) [reply Tasks/AzureResourceGroupDeploymentV2/task.json:17]: Ok, I just didn't know which week it was.

File: Tasks/AzureResourceManagerTemplateDeploymentV3/Tests/L0.ts
Comment (svrooij) [reply Tasks/AzureResourceManagerTemplateDeploymentV3/Tests/L0.ts:22]: One of the issues I had was that the tasks would not run on my local machine. `process.env["RUNNER_TEMP"]` would allow running the tests in Github Action (which I highly recommend for visibility), previously the Azure DevOps pipelines where not open to the world. I possibly have access now that my account is invited in the MS tenant.

`process.env["TMP"]` allows running on local machine, since all machines have `TMP` defined but probably not `AGENT_TEMPDIRECTORY`

File: Tasks/AzureResourceGroupDeploymentV2/Tests/L0.ts
Comment (svrooij) [reply Tasks/AzureResourceGroupDeploymentV2/Tests/L0.ts:12]: The tested used to fail locally due to this timeout, since this PR is pretty old I'm not sure if this change is still needed.

File: Tasks/AzureResourceGroupDeploymentV2/operations/ResourceGroup.ts
Comment (svrooij) [reply Tasks/AzureResourceGroupDeploymentV2/operations/ResourceGroup.ts:566]: the whole point of this PR was that it just outputs the actual value instead of a stringified version including quotes.

According to the [docs on variables](https://learn.microsoft.com/en-us/azure/devops/pipelines/process/set-variables-scripts?view=azure-devops&tabs=bash#about-tasksetvariable) it should output `##vso[task.setvariable variable=myVar;]foo` for a variable called `myVar` with a value `foo`. The current implementation (with stringify) outputs `##vso[task.setvariable variable=myVar;]"foo"` that means the value of that variable will be `"foo"`. And because it's `"foo"` it cannot be used in other steps and it's up to the user to parse the entire object.

The only time JSON.stringify is valid is for Arrays, then the value of the output should be `["value1", "value2"]` which is exactly what stringify does.

Comment (ozanovus) [reply Tasks/AzureResourceGroupDeploymentV2/operations/ResourceGroup.ts:566]: Oh; I see your point. However; since this quote behavior exists for a long time; there might be some users who are already use variables with `quotes`.

If we change behavior suddenly; these users might be affected. Let me discuss this topic with our PM.

Comment (svrooij) [reply Tasks/AzureResourceGroupDeploymentV2/operations/ResourceGroup.ts:566]: Mind that the old behavior **is not** in the current available documentation. I only found out by looking through the code. Probably no one is using it.

Comment (ozanovus) [reply Tasks/AzureResourceGroupDeploymentV2/operations/ResourceGroup.ts:566]: @svrooij; We've a conclusion for not disturbing existing behavior and your functionality.

Let's add a simple checkbox here:
image

With this explanation by default unchecking:
`[ ] Use value as it is without converting JSON`

If it is checked => it will use your change
if it is not checked => it will use existing mechanism

Do you know how to add a new input? Or we can also modify this PR accordingly if you're ok with changes in your source branch?

P.s. : we'll keep documentation's old version for "unchecked" behaviour and add your example as "checked" behavior.

Comment (svrooij) [reply Tasks/AzureResourceGroupDeploymentV2/operations/ResourceGroup.ts:566]: I have no clue on how to add inputs. I worked on this a long time ago 😄.

How about calling it `Stringify outputs using JSON.strinify (legacy)` and interpret not specified as `true`. Because it does something legacy which is considered incorrect. So `empty == true`, `true == true`, `false == false`. This could be checked with:

```js
if (inputs.legacyOutputs === false) {
// do my implementation
} else {
// do original (legacy) implementation
}
```

Then if people use the graphical interface the default value is unchecked and saved as such `false`. And people already using it can keep using the old behavior.

**Comment by svrooij** [comment]
What does this mean?

Of I add the checkbox you can accept the PR sooner? And of you have to do it it will take another 10 months?

**Comment by ozanaydinMSFT** [comment]
Related PR based on this suggestion is added with new checkbox:

https://github.com/microsoft/azure-pipelines-tasks/pull/18526

**Pull request closed without merging**
